flow/JsonTraceLogFormatter.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * JsonTraceLogFormatter.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/flow.h"
+#include "flow/JsonTraceLogFormatter.h"
+
+#include <sstream>
+
+void JsonTraceLogFormatter::addref() {
+	ReferenceCounted<JsonTraceLogFormatter>::addref();
+}
+
+void JsonTraceLogFormatter::delref() {
+	ReferenceCounted<JsonTraceLogFormatter>::delref();
+}
+
+const char* JsonTraceLogFormatter::getExtension() {
+	return "json";
+}
+
+const char* JsonTraceLogFormatter::getHeader() {
+	return "";
+}
+
+const char* JsonTraceLogFormatter::getFooter() {
+	return "";
+}
+
+namespace {
+
+void escapeString(std::stringstream& ss, const std::string source) {
+	for (auto c : source) {
+		if (c == '"') {
+			ss << "\\\"";
+		} else if (c == '\\') {
+			ss << "\\\\";
+		} else {
+			ss << c;
+		}
```

Comment:
One other thing I just noticed, it may be helpful to handle newlines (`\n` and `\r`) as well as the null byte (`\0`). The XML formatter converts all of those characters to spaces and even logs a warning for the null byte. 

I don't think we intentionally use any of these characters in our trace events, however there have been cases where they've be unintentionally used. For example, I think this might happen if someone tries to log a key or value without calling printable or something equivalent on it. If that were to happen, your logs would become invalid for newlines, and I think we had issues with data getting truncated with the null byte.

Reply:
We don't want to generate any invalid json, since that's a pain for anything consuming the logs. I propose we escape unprintable characters as valid json:

```
void escapeString(std::stringstream& ss, const std::string& source) {
    for (auto c : source) {
        if (c == '"') {
            ss << "\\\"";
        } else if (c == '\\') {
            ss << "\\\\";
        } else if (c == '\n') {
            ss << "\\n";
        } else if (c == '\r') {
            ss << "\\r";
        } else if (isprint(c)) {
            ss << c;
        } else {
            constexpr char hex[] = "0123456789abcdef";
            int x = int{static_cast<uint8_t>(c)};
            ss << "\\u00" << hex[x / 16] << hex[x % 16];
        }
    }
}
```

Reply:
That seems like a reasonable approach.